keps/sig-node/5328-node-declared-features/README.md

```diff
@@ -0,0 +1,995 @@
+
+The Node Declared Features framework addresses both issues through a unified mechanism without requiring any intervention (like adding Node Labels, selectors etc.). The scheduler uses the declared feature information to correctly filter nodes for new pods, while the admission controller validates operations against the declared features of a pod's current node. This ensures that feature incompatibilities are handled proactively at the control plane.
+
+## Design Details
+
+### API Changes
+
+Add a `DeclaredFeatures` field as type `[]string` to the `Node.Status` structure.
+
+```
+type NodeStatus struct {
+    // ... existing fields
+    // DeclaredFeatures provides a list of features declared by the node.
+    // The list is sorted alphabetically and contains no duplicate entries.
+    // +optional
+    // +featureGate=NodeDeclaredFeatures
+    DeclaredFeatures []string `json:"declaredFeatures,omitempty"`
+}
+
+
+// Node object remains unchanged in spec, only status is modified.
+type Node struct {
+    ...
+    Status NodeStatus `json:"status,omitempty"`
+}
+
+```
+
+**Note:**
+
+*   Any new feature being introduced in `node.status.declaredFeatures` is considered a formal API change and must go through the API review process. This governance will be enforced by protecting the list of declared features with `api-approvers` in the OWNERS file.
+*   We currently have [Node Features](https://github.com/kubernetes/api/blob/e8d4d542f6a9a16a694bfc8e3b8cd1557eecfc9d/core/v1/types.go#L6279) and [Node Runtime Features](https://github.com/kubernetes/api/blob/e8d4d542f6a9a16a694bfc8e3b8cd1557eecfc9d/core/v1/types.go#L6251) which publish some runtime features through Node Status. They are too narrowly scoped and currently not used for scheduling pods.
+
+
+#### Declared Feature Semantics
+
+1.  Combine Interdependent Settings
+    *   If multiple settings are required to enable a feature, they must be collapsed into a single, logical declared feature.
+    *   This simplifies decision-making for control plane components like the scheduler, which should not need to understand multiple interdependent settings.
+
+2.  Presence of a Declared Feature
+    *   A feature should only be present in `node.status.declaredFeatures` if it is enabled and functional.
+    *   If a feature is disabled or unsupported, its corresponding identifier must not exist in `node.status.declaredFeatures`.
+    *   This approach ensures consistency and simplifies logic for the control plane, treating nodes that don't know about a feature the same as those that have it disabled.
+
+3.  Validation Rules
+    *   Strings within the `node.status.declaredFeatures` slice must follow the standard identifier format used for most Kubernetes resource names (kebab-case with 63 character limit).
```

Comment:
I think we should use CamelCase, and generally match the name to the associated feature gate name. In cases where we need multiple declared features for a single feature gate (I'm not sure if there's a usecase for this yet), then I'd suggest a syntax like `FeatureName/SubFeature`

Reply:
We have some *long* feature gate names (`KubeletServiceAccountTokenForCredentialProviders`) and if we have the possibility of subfeatures `$GATE/...`, 63 might be a little tight. I guess *new* gates wanting to use this mechanism would be encouraged to be succinct, which is ~good.

If we want to give ourselves more wiggle-room, something like `const DNS1123SubdomainMaxLength int = 253` would also be ~ok as a limit.

We should also put an upper bound on the number of items in the list for sanity, something like 128 or 256 or something (how many in-progress feature rollouts do we think we'd ever have?)

Reply:
Modified to include camelCase and  DNS1123SubdomainMaxLength limit.  The validation would happen when a feature is being registered in the shared library. 

There was some discussion around upper bound [here]
(https://github.com/kubernetes/enhancements/pull/5347#discussion_r2145092605). I am not opposed to having an upper bound, but I am wondering what the behavior should be if we reach that bound. Since the library includes `minVersion` and `maxVersion` during feature registration, would it be a better approach be to have a reasonable default maxVersion (minVersion + X) ?. Say, If a feature does not graduate to GA X version's after it was introduced, we stop declaring it.